### DISCUSSION TITLE
$hashAlgorithm must be string, null used

### DISCUSSION THREAD

**Discussion by paul45** [codeigniter4/shield#313] [Q&A]
hello
i try shield in my project and i have this error

`CodeIgniter v4.2.1 Command Line Tool - Server Time:

An uncaught Exception was encountered

Type:        TypeError
Message:     Typed property CodeIgniter\Shield\Config\Auth::$hashAlgorithm must be string, null used
Filename:    C:\wamp64\codeigniter4\vendor\codeigniter4\framework\system\Config\Factories.php
Line Number: 126`

**Comment by datamweb** [comment]
Hello, what is the set value of `public string $hashAlgorithm`
 in file `app/Config/Auth.php`?
Please set 
```php
public string $hashAlgorithm = PASSWORD_DEFAULT;
```
If the problem is not solved, please provide more about the PHP version,Sheild version, how to install and etc?

**Comment by paul45** [comment]
yes $hashAlgorithm is PASSWORD_DEFAULT

 PHP: 7.4.0 — CodeIgniter: 4.2.1 
shield 1.0 beta

maybe it is a issue with codeigniter4/settings 2.1

**Comment by kenjis** [comment]
Install the latest `dev-develop` and run `php spark shield:setup`.
See https://github.com/codeigniter4/shield/blob/develop/docs/install.md#important-composer-error

**Reply by paul45** [reply]
I do this but the error is still there

**Reply by paul45** [reply]
i think the config file do not load correctly

**Reply by kenjis** [reply]
When?

**Reply by paul45** [reply]
TypeError

Typed property CodeIgniter\Shield\Config\Auth::$hashAlgorithm must be string, null used

SYSTEMPATH\Config\Factories.php at line 126

    119         }
    120 
    121         // Try to locate the class
    122         if (! $class = self::locateClass($options, $name)) {
    123             return null;
    124         }
    125 
    126         self::$instances[$options['component']][$class]    = new $class(...$arguments);
    127         self::$basenames[$options['component']][$basename] = $class;
    128 
    129         return self::$instances[$options['component']][$class];
    130     }
    131 
    132     /**
    133      * Finds a component class


    SYSTEMPATH\Common.php : 198   —  CodeIgniter\Config\Factories::__callStatic ( [arguments](http://localhost/login#) )

    191     /**
    192      * More simple way of getting config instances from Factories
    193      *
    194      * @return mixed
    195      */
    196     function config(string $name, bool $getShared = true)
    197     {
    198         return Factories::config($name, ['getShared' => $getShared]);
    199     }
    200 }
    201 
    202 if (! function_exists('cookie')) {
    203     /**
    204      * Simpler way to create a new Cookie instance.
    205      *

    VENDORPATH\codeigniter4\shield\src\Config\Services.php : 21   —   config()

    SYSTEMPATH\Config\BaseService.php : 251   —  CodeIgniter\Shield\Config\Services::auth ( [arguments](http://localhost/login#) )

    244     {
    245         $service = static::serviceExists($name);
    246 
    247         if ($service === null) {
    248             return null;
    249         }
    250 
    251         return $service::$name(...$arguments);
    252     }
    253 
    254     /**
    255      * Check if the requested service is defined and return the declaring
    256      * class. Return null if not found.
    257      */
    258     public static function serviceExists(string $name): ?string

**Reply by kenjis** [reply]
What did you do when the error occurred?

**Reply by paul45** [reply]
i have juste install Shield and the error occurred on each page before the Controllers file

**Reply by paul45** [reply]
before shield i use myth/Auth and i have remove the package

**Comment by kenjis** [comment]
Probably your PHP is too old.

> PASSWORD_DEFAULT was int 1; now is string '2y' (in PHP 7.4.0, 7.4.1, and 7.4.2 it was null) 
https://www.php.net/manual/en/migration74.incompatible.php#migration74.incompatible.core.password-algorithm-constants

Please try to check the value of `PASSWORD_DEFAULT`.

```console
bash-3.2$ php -v
PHP 7.4.30 (cli) (built: Jun  9 2022 09:30:10) ( NTS )
Copyright (c) The PHP Group
Zend Engine v3.4.0, Copyright (c) Zend Technologies
    with Xdebug v3.1.5, Copyright (c) 2002-2022, by Derick Rethans
    with Zend OPcache v7.4.30, Copyright (c), by Zend Technologies
```
```console
bash-3.2$ php -a
Interactive shell

php > var_dump(PASSWORD_DEFAULT);
php shell code:1:
string(2) "2y"
php > 
```

**Reply by paul45** [reply]
ok thx i have php 7.4.0 and PASSWORD_DEFAULT is null 

now i see my App/Config/Auth file isn't load only the file CodeIgniter\Shield\Config\Auth is load

**Reply by datamweb** [reply]
Can you post the output of command 
```console 
php spark namespaces
```
?

**Reply by paul45** [reply]
```
+-----------------------------+-------------------------------------------------------------+--------+
| Namespace                   | Path                                                        | Found? |
+-----------------------------+-------------------------------------------------------------+--------+
| CodeIgniter                 | C:\wamp64\codeigniter4\vendor\codeigniter4\framework\system | Yes    |
| App                         | C:\wamp64\codeigniter4\app                                  | Yes    |
| Config                      | C:\wamp64\codeigniter4\app\Config                           | Yes    |
| Commentaires                | C:\wamp64\codeigniter4\module\commentaires                  | Yes    |
| Commentaires\Tests\Support  | C:\wamp64\codeigniter4\module\commentaires\tests\_support   | Yes    |
+-----------------------------+-------------------------------------------------------------+--------+
```

**Reply by datamweb** [reply]
And output below, please?
```php 
<?php
namespace App\Controllers;

use App\Controllers\BaseController;
use Config\Auth;

class Home extends BaseController
{
    public function index()
    {

       $fc = new auth();
       dd($fc->hashAlgorithm);

    }
}
```

**Reply by paul45** [reply]
$fc->hashAlgorithm string (2) "2y"

⧉ Called from .../app/Controllers/Home.php:13 [dd()]

    .../codeigniter4/framework/system/CodeIgniter.php:896 [App\Controllers\Home->index()]
    .../codeigniter4/framework/system/CodeIgniter.php:466 [CodeIgniter\CodeIgniter->runController()]
    .../codeigniter4/framework/system/CodeIgniter.php:349 [CodeIgniter\CodeIgniter->handleRequest()]
    <ROOT>/index.php:55 [CodeIgniter\CodeIgniter->run()]

**Reply by paul45** [reply]
i have change 
    public string $hashAlgorithm = PASSWORD_DEFAULT;
to
    public string $hashAlgorithm = PASSWORD_BCRYPT;
in CodeIgniter\Shield\Config\Auth.php

because when i change in App\Config\Auth.php the value of hashAlgorithm  don't change

**Reply by datamweb** [reply]
Is file `App/Config/Auth.php` as below?
If yes, it shouldn't be a problem, I checked and by changing `public string $hashAlgorithm = PASSWORD_DEFAULT;` in file `App/Config/Auth.php`, 

> these changes are applied to the shield

### I will correct it, it does not apply.

```php
<?php

namespace Config;

use CodeIgniter\Shield\Config\Auth as ShieldAuth;
use CodeIgniter\Shield\Authentication\Actions\ActionInterface;
use CodeIgniter\Shield\Authentication\AuthenticatorInterface;
use CodeIgniter\Shield\Authentication\Authenticators\AccessTokens;
use CodeIgniter\Shield\Authentication\Authenticators\Session;
use CodeIgniter\Shield\Authentication\Passwords\ValidatorInterface;
use CodeIgniter\Shield\Models\UserModel;

class Auth extends ShieldAuth
{
```

**Reply by paul45** [reply]
yes the file is like this

**Reply by datamweb** [reply]
See #319 .
